libkrun/libkruntime.c

```diff
@@ -37,86 +79,254 @@ clock_gettime_monotonic()
 }
 
 /*
- * JNI Implementation -- Optionally compiled in
+ * JNI wrappers -- Optionally compiled in
  */
 #ifdef WITH_JAVA
+JNIEXPORT void JNICALL
+Java_IterationsRunner_JNI_1libkruntime_1init(JNIEnv *e, jclass c) {
+    libkruntime_init();
+}
+
+JNIEXPORT void JNICALL
+Java_IterationsRunner_JNI_1libkruntime_1done(JNIEnv *e, jclass c) {
+    libkruntime_done();
+}
 
 JNIEXPORT jdouble JNICALL
 Java_IterationsRunner_JNI_1clock_1gettime_1monotonic(JNIEnv *e, jclass c) {
     return (jdouble) clock_gettime_monotonic();
 }
 
 JNIEXPORT jlong JNICALL
-Java_IterationsRunner_JNI_1read_1ts_1reg_1start(JNIEnv *e, jclass c)
+Java_IterationsRunner_JNI_1read_1core_1cycles(JNIEnv *e, jclass c)
 {
-    return read_ts_reg_start();
+    return read_core_cycles();
 }
+#endif
 
-JNIEXPORT jlong JNICALL
-Java_IterationsRunner_JNI_1read_1ts_1reg_1stop(JNIEnv *e, jclass c)
+#if defined(__linux__) && !defined(TRAVIS)
+int
+open_msr_node(int core, int flags)
 {
-    return read_ts_reg_stop();
+    char path[MAX_MSR_PATH];
+    int msr_node;
+
+    /* Note this is not the default msr(4) device! */
+    if (snprintf(path, MAX_MSR_PATH, "/dev/cpu/%d/rmsr", core) < 0) {
+        perror("snprintf");
+        exit(EXIT_FAILURE);
+    }
+
+    msr_node = open(path, flags);
+    if (msr_node < 0) {
+        perror(path);
+        exit(EXIT_FAILURE);
+    }
+
+    return msr_node;
+}
+
+u_int64_t
+read_msr(int core, long addr)
+{
+    u_int64_t msr_val;
+    int msr_node;
+
+    msr_node = msr_r_nodes[core];
+
+    if (lseek(msr_node, addr, SEEK_SET) < 0) {
+        perror("lseek");
+        exit(EXIT_FAILURE);
+    }
+
+    if (read(msr_node, &msr_val, sizeof(msr_val)) != sizeof(msr_val)) {
+        perror("read");
+        exit(EXIT_FAILURE);
+    }
+
+    return msr_val;
+}
+
+void
+write_msr(int core, long addr, uint64_t msr_val)
+{
+    int msr_node;
+
+    msr_node = msr_w_nodes[core];
+
+    if (lseek(msr_node, addr, SEEK_SET) < 0) {
+        perror("lseek");
+        exit(EXIT_FAILURE);
+    }
+
+    if (write(msr_node, &msr_val, sizeof(msr_val)) != sizeof(msr_val)) {
+        perror("write");
+        exit(EXIT_FAILURE);
+    }
 }
-#endif
 
 /*
- * Support for reading the TS (timestamp) register.
- *
- * These functions assume an x86-64 CPU.
- *
- * For more info, see page 16 of:
- *
- * "How to Benchmark Code Execution Times on Intel® IA-32 and IA-64
- * Instruction Set Architectures" by Gabriele Paoloni.
+ * Configure fixed-function counter 1 to count all rings and threads
  */
+void
+config_fixed_ctr1(int core, int enable)
+{
+    u_int64_t msr_val = read_msr(core, MSR_IA32_FIXED_CTR_CTRL);
+    if (enable) {
+        msr_val |= (EN1_OS | EN1_USR | EN1_ANYTHR);
+    } else {
+        msr_val &= ~((u_int64_t) EN1_OS | EN1_USR | EN1_ANYTHR);
```

Comment:
Because those constants default to a signed 32.

In the first case, when you or `msr_val` (an unsigned 64) with the signed 32 (which is known to be +ve), the compiler will zero extend it, which is fine.

In the second case, we want to turn of all bits other than `EN1_OS | EN1_USR | EN1_ANYTHR`. This means we need the high bits (32-63) to be set in the mask. Without the cast, the 32-bit value would be zero extended and the top 32 bits would not be set.